src/garage/torch/q_functions/discrete_cnn_q_function.py

```diff
@@ -110,3 +111,21 @@ def __init__(self,
                          output_b_init=output_b_init,
                          layer_normalization=layer_normalization,
                          is_image=is_image)
+
+    # pylint: disable=arguments-differ
+    def forward(self, observations):
+        """Return Q-value(s).
+
+        Args:
+            observations (np.ndarray): observations of shape `input_shape`.
+
+        Returns:
+            torch.Tensor: Output value
+        """
+        if observations.shape != self._env_spec.observation_space.shape:
```

Comment:
just make this an error?

Reply:
I don't think this should be an error, this just unflattens a flattened batch of observations. CNNs specifically need them to be unflattened (while other primitives don't necessarily) so I think it makes sense to do this here.

Reply:
@krzentner can you weigh in? I think that it's always an error to pass an observation which doesn't match env_spec.observation_space.

Reply:
As I remember it, last time we discussed this we concluded the opposite: that `forward` would always take flattened observations, since it made everything except CNNs simpler. This logic should probably really live in akro eventually, but for the foreseeable future this is a more practical solution.